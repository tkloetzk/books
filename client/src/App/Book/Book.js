import React, { Component } from 'react';
import Card from '@material-ui/core/Card';
import CardHeader from '@material-ui/core/CardHeader';
import CardMedia from '@material-ui/core/CardMedia';
import CardContent from '@material-ui/core/CardContent';
import IconButton from '@material-ui/core/IconButton';
import Menu from '@material-ui/core/Menu';
import MenuItem from '@material-ui/core/MenuItem';
// import ReadBook from './read-book.svg';
import UnreadBook from '@material-ui/icons/BookOutlined';
import MoreVertIcon from '@material-ui/icons/MoreVert';
import { withStyles } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';

const styles = {
  card: {
    maxWidth: 200,
    margin: '6px',
    padding: '5px',
  },
  header: {
    '& span': {
      fontSize: '14px',
    },
    padding: '6px',
    height: '73px',
    textAlign: 'center',
  },
  content: {
    fontSize: '12px',
    padding: '6px',
    height: '145px',
  },
  media: {
    height: 0,
    paddingTop: '83%',
    backgroundSize: '100px',
  },
  actions: {
    display: 'flex',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  icon: {
    paddingLeft: '0px',
  },
};
function RatingDisplay(props) {
  if (!props.book) {
    return null;
  }

  if (props.book.adjustedRating) {
    return (
      <Typography>
        {Math.round(props.book.adjustedRating * 1000) / 1000}
      </Typography>
    );
  }

  return (
    <React.Fragment>
      <Typography style={{ textAlign: 'end' }}>
        Amazon Rating:{' '}
        {Math.round(props.book.amazonAverageRating * 1000) / 1000}
        <br />
        Goodreads Rating:{' '}
        {Math.round(props.book.goodreadsAverageRating * 1000) / 1000}
      </Typography>
    </React.Fragment>
  );
}
class Book extends Component {
  state = {
    anchorEl: null,
  };
  handleExpandClick = event => {
    console.log('clicked');
    this.setState({
      anchorEl: event.currentTarget,
    });
  };

  // TODO: Create modal instead of redirecting
  viewAmazonPage = href => {
    var win = window.open(href, '_blank');
    win.focus();
    this.handleClose();
  };
  handleClose = () => {
    this.setState({ anchorEl: null });
  };

  render() {
    const { book, classes } = this.props;
    const { anchorEl } = this.state;

    const title =
      book.title.length < 92 ? book.title : book.title.substring(0, 92) + '...';

    const subheader =
      book.subtitle.length < 45
        ? book.subtitle
        : book.subtitle.substring(0, 45) + '...';

    const description =
      book.description.length < 285
        ? book.description
        : book.description.substring(0, 285) + '...';

    const highlighted =
      !book.amazonAverageRating ||
      !book.amazonRatingsCount ||
      !book.goodreadsAverageRating ||
      !book.goodreadsRatingsCount;

    return (
      <Card
        className={classes.card}
        key={book.isbn}
        style={{ backgroundColor: highlighted ? 'yellow' : null }}
      >
        <CardHeader
          className={classes.header}
          action={
            book.href && (
              <IconButton onClick={this.handleExpandClick}>
                <MoreVertIcon />
              </IconButton>
            )
          }
          title={title}
          subheader={subheader}
        />
        <Menu
          id="simple-menu"
          anchorEl={anchorEl}
          open={Boolean(anchorEl)}
          onClose={this.handleClose}
        >
          <MenuItem onClick={() => this.viewAmazonPage(book.href)}>
            View on Amazon
          </MenuItem>
        </Menu>
        <CardMedia
          className={classes.media}
          image={book.thumbnail}
          title={book.title}
        />
        <CardContent className={classes.content}>{description}</CardContent>

        <div className={classes.actions}>
          <IconButton aria-label="Delete" className={classes.icon}>
            <UnreadBook fontSize="large" />
          </IconButton>
          <RatingDisplay book={book} />
        </div>
      </Card>
    );
  }
}

Book.defaultProps = {
  book: {
    title: '',
    subtitle: '',
    description: '',
  },
};
export default withStyles(styles)(Book);
