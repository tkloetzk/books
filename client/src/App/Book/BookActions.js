import React from 'react'
import Collapse from '@material-ui/core/Collapse';
import CardContent from '@material-ui/core/CardContent';
import EditableLabel from 'react-inline-editing';
import isEmpty from 'lodash/isEmpty';
import Typography from '@material-ui/core/Typography';

const BookActions = ({expanded = false, book, handleFocusOut}) => {
  const goodreadsAverageRating =
    Math.round(book.goodreadsAverageRating * 1000) / 1000;
  const amazonAverageRating =
     Math.round(book.amazonAverageRating * 1000) / 1000;

  return (
    <Collapse in={expanded} timeout="auto" unmountOnExit>
      <CardContent style={{ textAlign: 'center' }}>
        <Typography style={{ display: 'inline-flex' }} component="span">
          ISBN:{' '}
          <EditableLabel
            text={book.isbn}
            inputWidth="100px"
            inputHeight="25px"
            onFocus={() => {}}
            onFocusOut={text => handleFocusOut(text, 'isbn')}
          />
        </Typography>
        {!isEmpty(book.price) && (
          <Typography style={{ display: 'inline-flex' }} component="span">
            Price:{' '}
            <EditableLabel
              text={book.price}
              inputWidth="75px"
              inputHeight="25px"
              onFocus={() => {}}
              onFocusOut={text => handleFocusOut(text, 'price')}
            />
          </Typography>
        )}
        <Typography style={{ display: 'inline-flex' }} component="span">
          Amazon Rating:{' '}
          <EditableLabel
            text={`${amazonAverageRating}`}
            inputWidth="75px"
            inputHeight="25px"
            onFocus={() => {}}
            onFocusOut={text =>
              handleFocusOut(text, 'amazonAverageRating')
            }
          />
        </Typography>
        <Typography style={{ display: 'inline-flex' }} component="span">
          Goodreads Rating:{' '}
          <EditableLabel
            text={`${goodreadsAverageRating}`}
            inputWidth="75px"
            inputHeight="25px"
            onFocus={() => {}}
            onFocusOut={text =>
              handleFocusOut(text, 'goodreadsAverageRating')
            }
          />
        </Typography>
        <Typography style={{ display: 'inline-flex' }} component="span">
          Amazon Review:{' '}
          <EditableLabel
            text={`${book.amazonRatingsCount}`}
            inputWidth="75px"
            inputHeight="25px"
            onFocus={() => {}}
            onFocusOut={text =>
              handleFocusOut(text, 'amazonRatingsCount')
            }
          />
        </Typography>
        <Typography style={{ display: 'inline-flex' }} component="span">
          Goodreads Review:{' '}
          <EditableLabel
            text={`${book.goodreadsRatingsCount}`}
            inputWidth="75px"
            inputHeight="25px"
            onFocus={() => {}}
            onFocusOut={text =>
              handleFocusOut(text, 'goodreadsRatingsCount')
            }
          />
        </Typography>
        <Typography style={{ display: 'inline-flex' }} component="span">
          Category:{' '}
          <EditableLabel
            text={`${book.categories}`}
            inputWidth="75px"
            inputHeight="25px"
            onFocus={() => {}}
            onFocusOut={text => handleFocusOut(text, 'categories')}
          />
        </Typography>
      </CardContent>
    </Collapse>
  )
}

export default BookActions