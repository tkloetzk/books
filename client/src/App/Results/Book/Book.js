import ListItem from '@material-ui/core/ListItem';
import React, { Component } from 'react';
import { withStyles } from '@material-ui/core/styles';

const styles = {
  img: {
    maxWidth: '100%',
    height: 'auto',
    position: 'relative',
  },
};

class Book extends Component {
  render() {
    const { book, classes } = this.props;

    return (
      <ListItem>
        <div className="col-sm-1">
          <img src={book.image} className={`img-fluid ${classes.img}`} />
        </div>
      </ListItem>
    );
  }
}

export default withStyles(styles)(Book);
