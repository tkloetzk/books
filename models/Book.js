const mongoose = require('mongoose');
const Schema = mongoose.Schema;

const BookSchema = new Schema(
  {
    isbn: {
      type: String,
      required: true,
      unique: true,
    },
    id: {
      type: String,
      unique: true,
    },
    title: {
      type: String,
    },
    subtitle: {
      type: String,
    },
    thumbnail: {
      type: String,
    },
    href: {
      type: String,
    },
    price: {
      type: String,
    },
    categories: {
      type: Array,
    },
    description: {
      type: String,
    },
    amazonAverageRating: {
      type: Number,
    },
    goodreadsAverageRating: {
      type: Number,
    },
    amazonRatingsCount: {
      type: Number,
    },
    goodreadsRatingsCount: {
      type: Number,
    },
  },
  {
    collection: 'bookshelf',
  }
);

module.exports = Book = mongoose.model('book', BookSchema);
