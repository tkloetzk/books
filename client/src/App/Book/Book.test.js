import React from 'react';
import { shallow } from 'enzyme';
import { Book } from './Book';

describe('Book', () => {
  let props;
  let wrapper;
  let instance;

  beforeEach(() => {
    props = {
      handleSave: jest.fn(),
      book: {
        categories: ['Religion'],
        _id: '5c806393df78d5e84f8388ad',
        amazonAverageRating: 4.6,
        amazonRatingsCount: 20,
        price: '',
        isbn: '9780736917728',
        title: "Raising a Daughter After God's Own Heart",
        subtitle: '',
        description:
          'Elizabeth George, bestselling author and mother of two daughters, provides biblical insight and guidance for every mom who wants to lead their daughter to a godly life through example, study, and prayer. Elizabeth includes questions to draw moms and daughter closer as together they pursue spiritual priorities and God s heart."',
        thumbnail:
          'http://books.google.com/books/content?id=K_AhmQEACAAJ&printsec=frontcover&img=1&zoom=1&source=gbs_api',
        goodreadsAverageRating: 4.24,
        goodreadsRatingsCount: 148,
        __v: 0,
        adjustedRating: 4.225772626931567,
      },
    };
    wrapper = shallow(<Book {...props} />);
    instance = wrapper.instance();
  });
  afterEach(() => {
    jest.clearAllMocks();
  });
  describe('renders', () => {
    it('with no props', () => {
      wrapper = shallow(<Book />);
      expect(wrapper).toMatchSnapshot();
    });
    it('with correct props', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
