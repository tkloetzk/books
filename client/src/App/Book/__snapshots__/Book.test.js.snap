// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Book renders with correct props 1`] = `
<WithStyles(Card)
  className={null}
  key="9780736917728"
  style={
    Object {
      "margin": "6px",
      "maxHeight": 461,
      "maxWidth": 200,
      "padding": "5px",
    }
  }
>
  <div>
    <WithStyles(Typography)
      component="span"
      variant="body1"
    >
      <EditableLabel
        inputHeight="25px"
        inputWidth="190px"
        onFocus={[Function]}
        onFocusOut={[Function]}
        text="Raising a Daughter After God's Own Heart"
      />
    </WithStyles(Typography)>
    <WithStyles(Typography)
      component="span"
      variant="caption"
    >
      <EditableLabel
        inputHeight="25px"
        inputWidth="190px"
        onFocus={[Function]}
        onFocusOut={[Function]}
        text=""
      />
    </WithStyles(Typography)>
  </div>
  <WithStyles(CardMedia)
    image="http://books.google.com/books/content?id=K_AhmQEACAAJ&printsec=frontcover&img=1&zoom=1&source=gbs_api"
    title="Raising a Daughter After God's Own Heart"
  />
  <WithStyles(CardContent)>
    <EditableLabel
      inputHeight="25px"
      inputWidth="190px"
      onFocus={[Function]}
      onFocusOut={[Function]}
      text="Elizabeth George, bestselling author and mother of two daughters, provides biblical insight and guidance for every mom who wants to lead their daughter to a godly life through example, study, and prayer. Elizabeth includes questions to draw moms and daughter closer as together they pu..."
    />
  </WithStyles(CardContent)>
  <div>
    <WithStyles(IconButton)
      aria-label="Unread"
    >
      <pure(BookOutlinedIcon)
        fontSize="large"
      />
    </WithStyles(IconButton)>
    <WithStyles(IconButton)
      aria-expanded={false}
      aria-label="Show more"
      onClick={[Function]}
    >
      <pure(ExpandMoreIcon) />
    </WithStyles(IconButton)>
    <WithStyles(Typography)>
      4.226
    </WithStyles(Typography)>
  </div>
  <WithStyles(Collapse)
    in={false}
    timeout="auto"
    unmountOnExit={true}
  >
    <WithStyles(CardContent)
      style={
        Object {
          "textAlign": "center",
        }
      }
    >
      <WithStyles(Typography)
        component="span"
        style={
          Object {
            "display": "inline-flex",
          }
        }
      >
        ISBN:
         
        <EditableLabel
          inputHeight="25px"
          inputWidth="100px"
          onFocus={[Function]}
          onFocusOut={[Function]}
          text="9780736917728"
        />
      </WithStyles(Typography)>
      <WithStyles(Typography)
        component="span"
        style={
          Object {
            "display": "inline-flex",
          }
        }
      >
        Amazon Rating:
         
        <EditableLabel
          inputHeight="25px"
          inputWidth="75px"
          onFocus={[Function]}
          onFocusOut={[Function]}
          text="4.6"
        />
      </WithStyles(Typography)>
      <WithStyles(Typography)
        component="span"
        style={
          Object {
            "display": "inline-flex",
          }
        }
      >
        Goodreads Rating:
         
        <EditableLabel
          inputHeight="25px"
          inputWidth="75px"
          onFocus={[Function]}
          onFocusOut={[Function]}
          text="4.24"
        />
      </WithStyles(Typography)>
      <WithStyles(Typography)
        component="span"
        style={
          Object {
            "display": "inline-flex",
          }
        }
      >
        Amazon Review:
         
        <EditableLabel
          inputHeight="25px"
          inputWidth="75px"
          onFocus={[Function]}
          onFocusOut={[Function]}
          text="20"
        />
      </WithStyles(Typography)>
      <WithStyles(Typography)
        component="span"
        style={
          Object {
            "display": "inline-flex",
          }
        }
      >
        Goodreads Review:
         
        <EditableLabel
          inputHeight="25px"
          inputWidth="75px"
          onFocus={[Function]}
          onFocusOut={[Function]}
          text="148"
        />
      </WithStyles(Typography)>
      <WithStyles(Typography)
        component="span"
        style={
          Object {
            "display": "inline-flex",
          }
        }
      >
        Category:
         
        <EditableLabel
          inputHeight="25px"
          inputWidth="75px"
          onFocus={[Function]}
          onFocusOut={[Function]}
          text="Religion"
        />
      </WithStyles(Typography)>
    </WithStyles(CardContent)>
  </WithStyles(Collapse)>
</WithStyles(Card)>
`;

exports[`Book renders with no props 1`] = `""`;
